internal/tracer/middleware.go

```diff
@@ -39,12 +40,24 @@ func TracingRoundTripper(transport http.RoundTripper, collector Collector) http.
 		resp, err := transport.RoundTrip(req)
 		if err != nil {
 			builder.add(&ResponseError{Err: err})
-			cancel()
 			return nil, err
 		}
-		builder.add(&ResponseStart{Response: resp})
+
+		defer resp.Body.Close()
+		// Read the response body in full and then replace it with a new in-memory reader
+		// so that we can read the contents multiple times in any trace or middleware.
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Doesn't this break full-duplex bidi streams? I don't think you can do this.

What I was thinking is that instead, over in the reference client only (_not_ in the tracer), you'd have middleware that optionally captures the body. And it shouldn't do so by trying to read the entire response -- it needs to replace `Body` with a new implementation that will capture the bytes as they are read, like into a `*bytes.Buffer`. In fact, you could even store the `*bytes.Buffer` in the wire wrapper struct, and the middleware only needs to activate (to replace the body) when there is a wire wrapper in context.